File: src/2024/day05/2/main.go
package main

import (
	_ "embed"
	"fmt"
	"slices"
	"strconv"
	"strings"
)

//go:embed data.txt
var fileContents string

type rule struct {
	before, after int
}

func parse(input string) ([]rule, [][]int) {
	lines := strings.Split(strings.TrimSpace(input), "\n")

	separatorIdx := -1
	for i, line := range lines {
		if line == "" {
			separatorIdx = i
			break
		}
	}

	rules := make([]rule, len(lines[:separatorIdx]))
	for i, line := range lines[:separatorIdx] {
		parts := strings.Split(line, "|")
		before, _ := strconv.Atoi(parts[0])
		after, _ := strconv.Atoi(parts[1])
		rules[i] = rule{before, after}
	}

	// parse updates
	var updates [][]int
	for _, line := range lines[separatorIdx+1:] {
		var update []int
		for _, num := range strings.Split(line, ",") {
			n, _ := strconv.Atoi(num)
			update = append(update, n)
		}
		updates = append(updates, update)
	}

	return rules, updates
}

func isValidUpdate(update []int, rules []rule) bool {
	// for quick lookups
	indexMap := make(map[int]int)
	for i, num := range update {
		indexMap[num] = i
	}

	for _, rule := range rules {
		beforeIdx, beforeExists := indexMap[rule.before]
		afterIdx, afterExists := indexMap[rule.after]

		if beforeExists && afterExists && beforeIdx > afterIdx {
			return false
		}
	}
	return true
}

func buildGraph(update []int, rules []rule) map[int][]int {
	graph := make(map[int][]int)
	inDegree := make(map[int]int)

	for _, num := range update {
		if _, exists := graph[num]; !exists {
			graph[num] = []int{}
		}
	}

	for _, rule := range rules {
		if slices.Contains(update, rule.before) && slices.Contains(update, rule.after) {
			graph[rule.before] = append(graph[rule.before], rule.after)
			inDegree[rule.after]++
		}
	}

	return graph
}

func topologicalSort(update []int, rules []rule) []int {
	graph := buildGraph(update, rules)
	inDegree := make(map[int]int)
	result := make([]int, 0, len(update))

	// initial in-degrees
	for _, num := range update {
		inDegree[num] = 0
	}
	for _, rule := range rules {
		if slices.Contains(update, rule.before) && slices.Contains(update, rule.after) {
			inDegree[rule.after]++
		}
	}

	// find nodes with no incoming edges
	var queue []int
	for _, num := range update {
		if inDegree[num] == 0 {
			queue = append(queue, num)
		}
	}

	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]
		result = append(result, current)

		for _, neighbor := range graph[current] {
			inDegree[neighbor]--
			if inDegree[neighbor] == 0 {
				queue = append(queue, neighbor)
			}
		}
	}

	return result
}

func main() {
	rules, updates := parse(fileContents)

	sum := 0
	for _, update := range updates {
		if !isValidUpdate(update, rules) {
			sortedUpdate := topologicalSort(update, rules)
			middleIdx := len(sortedUpdate) / 2
			sum += sortedUpdate[middleIdx]
		}
	}

	fmt.Println(sum)
}

File: src/2024/day05/2/main.rs
use std::collections::{HashMap, VecDeque};
use std::fs;
use std::path::Path;

fn read_lines() -> Vec<String> {
    let current_file = Path::new(file!());
    let current_dir = current_file
        .parent()
        .expect("Failed to get parent directory");
    let file_path = current_dir.join("data.txt");

    fs::read_to_string(file_path)
        .expect("Failed to read file")
        .lines()
        .map(|line| line.to_string())
        .collect()
}

#[derive(Debug)]
struct Rule {
    before: i32,
    after: i32,
}

fn parse_input(input: &str) -> (Vec<Rule>, Vec<Vec<i32>>) {
    let mut sections = input.trim().split("\n\n");

    let rules = sections
        .next()
        .unwrap()
        .lines()
        .map(|line| {
            let mut parts = line.split('|');
            Rule {
                before: parts.next().unwrap().parse().unwrap(),
                after: parts.next().unwrap().parse().unwrap(),
            }
        })
        .collect();

    let updates = sections
        .next()
        .unwrap()
        .lines()
        .map(|line| line.split(',').map(|n| n.parse().unwrap()).collect())
        .collect();

    (rules, updates)
}

fn is_valid_update(update: &[i32], rules: &[Rule]) -> bool {
    let index_map: HashMap<i32, usize> = update.iter().enumerate().map(|(i, &n)| (n, i)).collect();

    rules.iter().all(
        |rule| match (index_map.get(&rule.before), index_map.get(&rule.after)) {
            (Some(&before_idx), Some(&after_idx)) => before_idx < after_idx,
            _ => true,
        },
    )
}

fn topological_sort(update: &[i32], rules: &[Rule]) -> Vec<i32> {
    let mut graph: HashMap<i32, Vec<i32>> = HashMap::new();
    let mut in_degree: HashMap<i32, i32> = HashMap::new();

    update.iter().for_each(|&n| {
        graph.entry(n).or_default();
        in_degree.insert(n, 0);
    });

    // graph building
    for rule in rules {
        if update.contains(&rule.before) && update.contains(&rule.after) {
            graph.entry(rule.before).or_default().push(rule.after);
            *in_degree.entry(rule.after).or_default() += 1;
        }
    }

    let mut result = Vec::new();
    let mut queue: VecDeque<i32> = update
        .iter()
        .filter(|&&n| in_degree[&n] == 0)
        .copied()
        .collect();

    while let Some(current) = queue.pop_front() {
        result.push(current);

        if let Some(neighbors) = graph.get(&current) {
            for &next in neighbors {
                in_degree.entry(next).and_modify(|e| *e -= 1);
                if in_degree[&next] == 0 {
                    queue.push_back(next);
                }
            }
        }
    }

    result
}

fn main() {
    let lines = read_lines();
    let content = lines.join("\n");
    let (rules, updates) = parse_input(&content);

    let sum_invalid = updates
        .iter()
        .filter(|update| !is_valid_update(update, &rules))
        .map(|update| {
            let sorted = topological_sort(update, &rules);
            sorted[sorted.len() / 2]
        })
        .sum::<i32>();

    println!("{}", sum_invalid);
}
